### PR TITLE
Fix stream state machine

### LINKED ISSUE #572: Should we transition to Open state when recv even MAX_STREAM_DATA or RST on idle
While thinking about implementation of the handling of idle state, it seems like we'd practically allocate the stream state object anyway if we get a MAX_STREAM_DATA or STREAM_BLOCKED (to keep track of flow control).

It seems better then to treat these streams as equivalent to Open streams because it'd simplify the logic of deciding the Stream limit count to advertise. If the reason we want to allow this transition is to deal with lost packets, we'd practically allocate the stream data eventually when the packets are recovered. 

Proposing a change where:

Idle -> Open

transition occurs not only on STREAM, but also BLOCKED, MAX_STREAM_DATA as well.

### PR DESCRIPTION
**Pull request by martinthomson**
This corrects a few infidelities about the way that stream states
update.  This includes Subodh's excellent suggestion to limit
transitions from idle to only go through open.

Fixes #572, #571.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2428,18 +2417,28 @@ A stream in the "open" state may be used by both peers to send frames of any
...
+From the "open" state, either endpoint can send a frame with the FIN flag set,
+which causes the stream to transition into one of the "half-closed" states.
+This flag can be set on the frame that opens the stream, which causes the stream
+to immediately become "half-closed".  Once an endpoint has completed sending all
+stream data and a STREAM frame with a FIN flag, including any retransmissions,
+the stream state becomes "half-closed (local)".  When an endpoint receives all
+stream data a FIN flag the stream state becomes "half-closed (remote)".  An
 endpoint MUST NOT consider the stream state to have changed until all data has
-arrived.
+either been acknowledged or received.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: I'm not sure if it is obvious that a RST is also considered acknowledged data here.

Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: But from the following it seems this also not required.

Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Thanks for picking up on that.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2428,18 +2417,28 @@ A stream in the "open" state may be used by both peers to send frames of any
...
+From the "open" state, either endpoint can send a frame with the FIN flag set,
+which causes the stream to transition into one of the "half-closed" states.
+This flag can be set on the frame that opens the stream, which causes the stream
+to immediately become "half-closed".  Once an endpoint has completed sending all
+stream data and a STREAM frame with a FIN flag, including any retransmissions,
+the stream state becomes "half-closed (local)".  When an endpoint receives all
+stream data a FIN flag the stream state becomes "half-closed (remote)".  An
 endpoint MUST NOT consider the stream state to have changed until all data has
-arrived.
+been acknowledged, received or discarded.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: The "arrived" was deliberate -- the stream does not need to hang around in open state until all data is acked. That's a matter of reliability, which should not have anything to do with stream states. I would like to keep stream states decoupled from reliability requirements.

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I would suggest that you revert the changes that require receipt of acks for sent data for streams to transition to closed. The other changes LGTM.
